### PR TITLE
Remove unnecessary logic in ProcessorTask::finalizeIfNecessary

### PR DESCRIPTION
**Pull request by ted-wq-x**
# Description

The `Sink::initLocalState` method is only useful when a query starts with `optional match`,we can move this logic into the `ResultCollector::finalizeInternal` method to optimize performance.

## Performance test

In sf100 datasets,query `match p=(a:person {id:24189255912498})-[f]-(b) return p limit 20`,10 concurrency, QPS improve 8.2%.

image

# Contributor agreement

- [x] I have read and agree to the [Contributor Agreement](https://github.com/kuzudb/kuzu/blob/master/CLA.md).

### CODE REVIEW

File: src/processor/operator/persistent/node_batch_insert.cpp [hunk 1]
```diff
@@ -207,19 +219,17 @@ void NodeBatchInsert::appendIncompleteNodeGroup(transaction::Transaction* transa
...
+                context->clientContext->getMemoryManager(), errorHandler);
         }
     }
     if (nodeSharedState->globalIndexBuilder) {
-        auto* localNodeState = ku_dynamic_cast<NodeBatchInsertLocalState*>(localState.get());
-        KU_ASSERT(localNodeState->errorHandler.has_value());
-        nodeSharedState->globalIndexBuilder->finalize(context,
-            localNodeState->errorHandler.value());
-        localNodeState->errorHandler->flushStoredErrors();
+        nodeSharedState->globalIndexBuilder->finalize(context, errorHandler);
```
Comment (royi-luo) [inline src/processor/operator/persistent/node_batch_insert.cpp:231]: Just to help guard against misuse can you add:
```suggestion
        KU_ASSERT(!localNodeState->errorHandler.has_value())
```

File: src/include/processor/processor_task.h [hunk 2]
```diff
@@ -15,7 +15,6 @@ class ProcessorTask : public common::Task {
     void run() override;
     void finalizeIfNecessary() override;
 
-private:
     static std::unique_ptr<ResultSet> populateResultSet(Sink* op,
         storage::MemoryManager* memoryManager);
```
Comment (royi-luo) [inline src/include/processor/processor_task.h:19]: can you add back the `private` below this function so that the members are still private?

**Review by royi-luo — APPROVED** [review]
Node batch insert changes look good to me, thanks! It'd be helpful next time if you included a bit more detail in the PR description (in this case why the parts in `finalizeIfNecessary` were unnecessary) just to save your reviewers some time spent understanding the changes.

**Comment by ted-wq-x** [comment]
Done

File: src/include/processor/processor_task.h (hunk 2 above)
Comment (ted-wq-x) [reply src/include/processor/processor_task.h:19]: There is a `private` below.

**Comment by royi-luo** [comment]
@ted-wq-x the CI checks look to be working so can you fix the errors (you can ignore the clang format job but the other ones should hopefully pass)

**Comment by ted-wq-x** [comment]
@royi-luo CI Workflow requires approval

**Comment by royi-luo** [comment]
> @royi-luo CI Workflow requires approval

Just approved, I guess it is pretty inconvenient for you to need approval every time to run CI, sorry about that.

**Review by royi-luo — APPROVED** [review]

**Comment by ted-wq-x** [comment]
Is there any problem about this pr?

**Comment by royi-luo** [comment]
> Is there any problem about this pr?

From the extension test failures it looks like we also need to update the finalize in `json_scan.cpp`. Other than that it looks good to me. I can take this over and make the remaining changes myself, it'll probably be easier for me to run CI too. Thanks!

**Pull request merged**